core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -0,0 +1,565 @@
+   private volatile int valuesTopology = -1;
+   private Object valuesLock = new Object();
+
+   @Inject
+   public void init(Configuration configuration,
+                    @ComponentName(ASYNC_TRANSPORT_EXECUTOR) ExecutorService executorService,
+                    CommandsFactory commandsFactory,
+                    RpcManager rpcManager,
+                    DataContainer<K, ?> dataContainer,
+                    PersistenceManager persistenceManager,
+                    StateConsumer stateConsumer) {
+      this.configuration = configuration;
+      this.executorService = executorService;
+      this.commandsFactory = commandsFactory;
+      this.rpcManager = rpcManager;
+      this.dataContainer = dataContainer;
+      this.persistenceManager = persistenceManager;
+      this.stateConsumer = stateConsumer;
+   }
+
+   @Start(priority = 15) // before StateConsumerImpl and StateTransferManagerImpl
+   public void start() {
+      // this assumes that number of segments does not change
+      numSegments = configuration.clustering().hash().numSegments();
+      stateTransferVersions = new AtomicLongArray(numSegments);
+      segmentVersions = new AtomicLongArray(numSegments);
+      // TODO: after preload, the entries can already have some versions. Update based on these
+      segmentStates = new AtomicReferenceArray<>(numSegments);
+      CacheTopology cacheTopology = stateConsumer.getCacheTopology();
+      ConsistentHash consistentHash = cacheTopology == null ? null : cacheTopology.getReadConsistentHash();
+      for (int i = 0; i < numSegments; ++i) {
+         // The component can be rewired, and then this is executed without any topology change
+         SegmentState state = SegmentState.NOT_OWNED;
+         if (consistentHash != null && consistentHash.isSegmentLocalToNode(rpcManager.getAddress(), i)) {
+            state = SegmentState.OWNED;
+         }
+         segmentStates.set(i, state);
+         log.tracef("Node %s, segment %d, setting state to %s", rpcManager.getAddress(), i, state);
+      }
+      syncRpcOptions = rpcManager.getRpcOptionsBuilder(ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS, DeliverOrder.NONE).build();
+      scheduledKeys = new ConcurrentHashMap<>(invalidationBatchSize);
+      invalidationBatchSize = configuration.clustering().invalidationBatchSize();
+      removedKeys = new ConcurrentHashMap<>(invalidationBatchSize);
+   }
+
+   /*@Start(priority = 61)
+   public void initSegmentVersions() {
+      // This is a forced preload but only after we have cache topology installed (and we can check segments, therefore)
+      CacheTopology cacheTopology = stateConsumer.getCacheTopology();
+      // during start from state, topology can be null here
+      ConsistentHash ch = cacheTopology.getWriteConsistentHash();
+      persistenceManager.processOnAllStores(new WithinThreadExecutor(), KeyFilter.ACCEPT_ALL_FILTER,
+         (marshalledEntry, taskContext) -> {
```

Comment:
Hmm, this is actually something not implemented properly. Versions are not updated when loading the data from cache store, and these are not persisted either. I should put a stick note to finish this.